### PR TITLE
Implemented SEO category functionality on storefront

### PR DESCRIPTION
**Pull request by sebaholesz**
| Q             | A
| ------------- | ---
|Description, reason for the PR| In the recent past, the SEO categories functionality was removed while refactoring query filters and parameters. After this refactoring, we were now able to re-introduce the functionality on storefront. This PR not only does that, but also brings minor changes to the API, such as "SEO insensitive" filters, which don't break the SEO category. In the past, these only included brands, but now also price and slider filters keep the user in the SEO category. 
|New feature| Yes
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: project-base/storefront/hooks/useQueryParams.ts [hunk 1]
```diff
@@ -175,22 +270,35 @@ export const useQueryParams = () => {
...
 
-    const pushQueries = (queries: UrlQueries, isPush?: boolean) => {
+    const pushQueries = (queries: UrlQueries, isPush?: boolean, pathnameOverride?: string) => {
         // remove queries which are not set or removed
         const filteredQueries = getFilteredQueries(queries);
 
         router[isPush ? 'push' : 'replace'](
             {
-                pathname: router.asPath.split('?')[0],
+                pathname: pathnameOverride ? pathnameOverride : router.asPath.split('?')[0],
```
Comment (tvikito) [inline project-base/storefront/hooks/useQueryParams.ts:301]:
```suggestion
                pathname: pathnameOverride || router.asPath.split('?')[0]
```

File: project-base/storefront/pages/categories/[categorySlug].tsx [hunk 2]
```diff
@@ -30,55 +32,61 @@ import {
...
+    useHandleDefaultFiltersUpdate(categoryData?.products);
+    useHandleSeoCategorySlugUpdate(categoryData);
 
     const seoTitle = useSeoTitleWithPagination(
-        categoryData?.category?.products.totalCount,
-        categoryData?.category?.name,
-        categoryData?.category?.seoTitle,
+        categoryData?.products.totalCount,
+        categoryData?.name,
+        categoryData?.seoTitle,
```
Comment (tvikito) [inline project-base/storefront/pages/categories/[categorySlug].tsx:61]: I'm loving it! 💯

File: project-base/storefront/pages/categories/[categorySlug].tsx [hunk 3]
```diff
@@ -132,3 +146,37 @@ export const getServerSideProps = getServerSidePropsWithRedisClient((redisClient
 });
 
 export default CategoryDetailPage;
+
+const useCategoryDetailData = (filter: ProductFilterApi | null): [undefined | CategoryDetailFragmentApi, boolean] => {
+    const client = useClient();
+    const router = useRouter();
+    const t = useTypedTranslationFunction();
+    const urlSlug = getSlugFromUrl(getUrlWithoutGetParameters(router.asPath));
+    const orderingMode = getProductListSort(parseProductListSortFromQuery(router.query[SORT_QUERY_PARAMETER_NAME]));
```
Comment (tvikito) [inline project-base/storefront/pages/categories/[categorySlug].tsx:155]: Another place where you can take `sort` value from `useQueryParams`

**Review by tvikito — CHANGES_REQUESTED** [review]
Very nice! I am surprised how relatively clean this implementation is 🎉

File: project-base/storefront/components/Pages/CategoryDetail/CategoryDetailContent.tsx [hunk 4]
```diff
@@ -15,7 +15,6 @@ import { useTypedTranslationFunction } from 'hooks/typescript/useTypedTranslatio
 import { useRouter } from 'next/router';
 import { useCallback, useRef, useState } from 'react';
 import { twJoin } from 'tailwind-merge';
-
```
Comment (tvikito) [inline project-base/storefront/components/Pages/CategoryDetail/CategoryDetailContent.tsx:18]: Just a huge nitpick but let's put the space here back.

File: project-base/storefront/pages/categories/[categorySlug].tsx (hunk 3 above)
Comment (tvikito) [reply project-base/storefront/pages/categories/[categorySlug].tsx]: I don't see the change...

**Review by tvikito — APPROVED** [review]

File: project-base/storefront/components/Pages/CategoryDetail/CategoryDetailContent.tsx (hunk 4 above)
Comment (sebaholesz) [reply project-base/storefront/components/Pages/CategoryDetail/CategoryDetailContent.tsx]: Done

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php [hunk 5]
```diff
@@ -146,9 +156,19 @@ function ($offset, $limit) use ($argument, $productFilterData, $flag, $batchLoad
...
+    {
+        throw new DeprecatedMethodException();
+    }
+
+    /**
+     * @param \Overblog\GraphQLBundle\Definition\Argument $argument
+     * @param \GraphQL\Type\Definition\ResolveInfo $info
+     * @return \App\FrontendApi\Model\Product\Connection\ProductExtendedConnection|\GraphQL\Executor\Promise\Promise
+     */
+    public function productsWithOverlyingEntityQuery(Argument $argument, ResolveInfo $info)
```
Comment (TomasLudvik) [inline project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:171]:
```suggestion
    public function productsWithOverlayingEntityQuery(Argument $argument, ResolveInfo $info)
```

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php [hunk 6]
```diff
@@ -160,6 +180,34 @@ public function productsQuery(Argument $argument)
             $argument,
         );
 
+        if ($argument['categorySlug']) {
```
Comment (TomasLudvik) [inline project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:183]:
```suggestion
        if ($argument['categorySlug'] !== null) {
```

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php [hunk 7]
```diff
@@ -160,6 +180,34 @@ public function productsQuery(Argument $argument)
...
+            $category = $this->categoryQuery->categoryOrSeoMixByUuidOrUrlSlugQuery($info, urlSlug: $argument['categorySlug']);
```
Comment (TomasLudvik) [inline project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:184]:
```suggestion
            $category = $this->categoryQuery->categoryOrSeoMixByUuidOrUrlSlugQuery($info, $argument['categorySlug']);
```

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php [hunk 8]
```diff
@@ -160,6 +180,34 @@ public function productsQuery(Argument $argument)
...
+
+            return $this->productsByCategoryOrReadyCategorySeoMixQuery(
+                $argument,
+                $category,
+            );
+        }
+
+        if ($argument['brandSlug']) {
```
Comment (TomasLudvik) [inline project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:192]:
```suggestion
        if ($argument['brandSlug'] !== null) {
```

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php [hunk 9]
```diff
@@ -160,6 +180,34 @@ public function productsQuery(Argument $argument)
...
+            /** @var \App\Model\Product\Brand\Brand $brand */
+            $brand = $this->brandQuery->brandByUuidOrUrlSlugQuery(urlSlug: $argument['brandSlug']);
```
Comment (TomasLudvik) [inline project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:194]:
```suggestion
            $brand = $this->brandQuery->brandByUuidOrUrlSlugQuery($argument['brandSlug']);
```

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php [hunk 10]
```diff
@@ -160,6 +180,34 @@ public function productsQuery(Argument $argument)
...
+
+            return $this->productsByBrandQuery(
+                $argument,
+                $brand,
+            );
+        }
+
+        if ($argument['flagSlug']) {
```
Comment (TomasLudvik) [inline project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:202]:
```suggestion
        if ($argument['flagSlug'] !== null) {
```

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php [hunk 11]
```diff
@@ -160,6 +180,34 @@ public function productsQuery(Argument $argument)
...
+            $flag = $this->flagQuery->flagByUuidOrUrlSlugQuery(urlSlug: $argument['flagSlug']);
```
Comment (TomasLudvik) [inline project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:203]:
```suggestion
            $flag = $this->flagQuery->flagByUuidOrUrlSlugQuery($argument['flagSlug']);
```

**Review by TomasLudvik — CHANGES_REQUESTED** [review]

File: project-base/app/src/FrontendApi/Component/SessionChecker/SessionChecker.php [hunk 12]
```diff
@@ -15,7 +15,7 @@ public function onKernelResponse(ResponseEvent $event): void
     {
         $request = $event->getRequest();
 
-        if (!$request->hasSession() || !$request->getSession()->isStarted() || !str_contains($request->getRequestUri(), 'graphql')) {
+        if (!$request->hasSession() || !$request->getSession()->isStarted() || !str_contains($request->getRequestUri(), '/graphql/')) {
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply project-base/app/src/FrontendApi/Component/SessionChecker/SessionChecker.php:18]: GraphQL panel in the profiler does not work (http://127.0.0.1:8000/_profiler/xxxxxx?panel=graphql), due to this condition. 
This change was not necessary for this feature, but I have fixed it because I needed to check some graphql queries

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 7 above)
Comment (grossmannmartin) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:184]: second argument is UUID, not slug...

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 9 above)
Comment (grossmannmartin) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:194]: first argument is uuid, not slug

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 11 above)
Comment (grossmannmartin) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:203]: first argument is uuid, not slug

File: project-base/app/src/Component/Router/FriendlyUrl/FriendlyUrlFacade.php [hunk 13]
```diff
@@ -105,6 +105,17 @@ public function getUndefinedRouteNamesInMapping(): array
...
 
+    /**
+     * @param string $routeName
+     * @return string
+     */
+    public function getEntityClassByRouteName(string $routeName): string
+    {
+        $routeNameMapping = $this->friendlyUrlRepository->getRouteNameToEntityMap();
+
+        return $routeNameMapping[$routeName] ?? throw new FriendlyUrlNotFoundException();
```
[the remark this answers is not in the document]
Comment (sebaholesz) [reply project-base/app/src/Component/Router/FriendlyUrl/FriendlyUrlFacade.php:116]: Done

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 8 above)
Comment (sebaholesz) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:192]: Done

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 10 above)
Comment (sebaholesz) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:202]: Done

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 5 above)
Comment (sebaholesz) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:171]: Here I would guide the decision on this explanation:

_The terms "overlying" and "overlaying" are both related to the concept of something being on top of something else, but they are used in slightly different contexts.

    "Overlying": This term is often used in a more passive sense to describe something that is located above or on top of something else. It is often used in geological or medical contexts. For example, in geology, you might say that a layer of sandstone is "overlying" a layer of shale, meaning it is positioned above the shale in the strata. In medicine, you might refer to skin "overlying" a bone or muscle, meaning it covers it.

    "Overlaying": This term is often used in a more active sense to describe the act of placing something on top of something else. It is often used in the context of design, graphics, and technology. For example, in graphic design, you might "overlay" text on an image, meaning you actively place the text over the image. In technology, an "overlay" might refer to a user interface element that appears on top of other content on a screen.

So, in summary, "overlying" often describes a state of being or position, while "overlaying" often describes an action or process._

Because of that, I think overlying is more suitable

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 7 above)
Comment (sebaholesz) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:184]: After discussing the changes, I removed the named parameter and included `null` for the UUID.

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 9 above)
Comment (sebaholesz) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:194]: After discussing the changes, I removed the named parameter and included `null` for the UUID.

File: project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php (hunk 11 above)
Comment (sebaholesz) [reply project-base/app/src/FrontendApi/Resolver/Products/ProductsQuery.php:203]: After discussing the changes, I removed the named parameter and included `null` for the UUID.

**Review by TomasLudvik — APPROVED** [review]
